### ISSUE TITLE
Feedback

### ISSUE THREAD

**Issue by whyrusleeping** [hsanjuan/ipfsclusterspec#1]
Hey @hsanjuan some great ideas here, I thought i'd provide my thoughts.

First off, i'm not super sold on using raft. In my experience it takes a decent chunk of extra work to get it to scale past ten or so nodes. (Things may have changed since i last looked into it though). I would leave the consensus method itself variable, maybe we can start off using raft, but plugging in good old paxos instead should be easy to do.

On replication, you mention that its needed at the top but don't go into any detail on the specifics of how that should be decided on. How do we decide who should be storing what block? Also, a 'replication level' should probably be specified in the global log, so that we can change replication on the fly, ensure all nodes know about it, and proceed to redistribute blocks.

It the status checks, nodes should be able to provide information about whether or not they are 'up to date' on replicating the blocks assigned to them.

**Comment by hsanjuan** [comment]
First of all, thanks for the feedback @whyrusleeping. I am very much in line with your comments:

> First off, i'm not super sold on using raft. In my experience it takes a decent chunk of extra work to get it to scale past ten or so nodes. (Things may have changed since i last looked into it though). I would leave the consensus method itself variable, maybe we can start off using raft, but plugging in good old paxos instead should be easy to do.

This is true and I have to agree. I raised the question that in this initial design scalability would depend on RAFT capacity. To go further, the main problem is that RAFT or even Paxos are not designed to scale into many participants (consensus is noisy and has lots of overhead with many participants). The way to use them is to have a small cluster which holds distributed-but-small coordination information and pools of workers that do not participate in decision-making but can get things done and scale.

So a further step in IPFSCluster would be that "IPFSCluster nodes" (those involved in consensus) do not perform the heavy work but rather have a pool of workers to do it (monitoring of IPFS nodes in the wide sense, checking that things are correctly persisted etc). To be honest, I was more worried about the limitations that IPNS would put on the system if it was used to do  the message-passing. If that is not a constraint and messages can be passed efficiently, orchestrating pools of workers which are not part of the consensus protocol but perform tasks and report back to their queen becomes a way easier and plausible task.

> On replication, you mention that its needed at the top but don't go into any detail on the specifics of how that should be decided on.  How do we decide who should be storing what block?

Yep, multiple strategies possible. For example, free disk space. But also, bandwidth, physical collocation (do not store 2 copies in the same rack), read/write performance... This should be a separate "plugin" too. Something like that Cassandra "snitches" (https://docs.datastax.com/en/cassandra/2.0/cassandra/architecture/architectureSnitchesAbout_c.html) that allows to choose the strategy that works better for the environment you're deploying into. Extending IPFS nodes to provide a /health or /stat endpoint would come handy here.

> Also, a 'replication level' should probably be specified in the global log, so that we can change replication on the fly, ensure all nodes know about it, and proceed to redistribute blocks.

In fact, I was thinking it is cheap to include `replication_factor` per object (apart from a global minimum one). It would be a key in the `assignments` sections of the global log and then part of the `hash_list` file for every hash. Replication factor would be like any other metadata that can be included in the same way.

> It the status checks, nodes should be able to provide information about whether or not they are 'up to date' on replicating the blocks assigned to them.

Yes, actually is a complex problem because it's not only knowing how far a node is with a file, but also detecting it is not stuck or that it's not terribly slow, or that it got full because some other that was ingested on the side. To be fine-grained here probably requires having bitswap expose info on how transfers are going. The dummy strategy is to just wait for things to appear in the "havelist" until you have waited for too long  (say, by a conservative estimate based on the file size) and you pick a different node to do it.

I'll try to include some of these thoughts on the Readme in the following days.

**Comment by whyrusleeping** [comment]
Yeah, in my experience, raft doesnt really scale past ten or so nodes. Paxos i've seen hit over three hundred pretty readily. Raft is much simpler, and has nice implementations we can quickly grab and use.  The golang (it doesnt necessarily have to be go, but thats my preference) libs for paxos are less exciting and confidence inspiring. I like your idea about 'sub clusters' and thats a similar conclusion to the idea a friend of mine was discussing as a way to make raft scale. Basically having layers of consensus. That gets complicated far more quickly though, and thats definitely a 'down the road' type of thing.

If we're going to do consensus for this, i think it should be done in a side channel. Using IPNS for that would be cool and all, but i don't think its currently a good fit for the task (we can and are improving it, but not at a rate where it will be _good_ for this task by the time we want to implement this). We should either use regular old tcp (boring) or make it into a libp2p service (super cool!) which basically acts like a tcp socket, but we get NAT traversal and stream multiplexing and all that jazz.

:+1: on making replication pluggable. Although lets be careful about attaching metadata per hash, that might get to be a significant overhead. I think adding metadata per 'pinset' might be a better tactic there.

Keeping tabs on status is just going to be really hard. sooooo many things can go wrong. Adding some API to bitswap to query transfer progress will be a good thing to help this out (and its something people have wanted for a while).

Anyways, great work and I look forward to seeing further revisions :)

**Comment by hsanjuan** [comment]
> Yeah, in my experience, raft doesnt really scale past ten or so nodes. Paxos i've seen hit over three hundred pretty readily. 

Cool, where can I see this?

> The golang (it doesnt necessarily have to be go, but thats my preference)

Yes, I'd use Go.

> If we're going to do consensus for this, i think it should be done in a side channel. Using IPNS for that would be cool and all, but i don't think its currently a good fit for the task (we can and are improving it, but not at a rate where it will be good for this task by the time we want to implement this). We should either use regular old tcp (boring) or make it into a libp2p service (super cool!) which basically acts like a tcp socket, but we get NAT traversal and stream multiplexing and all that jazz.

Well, my idea was that it had to be IPFS-based (so it can run with no other requirements than those of IPFS). If we have boring tcp, then we bring up a Zookeeper cluster directly for coordination. And then you don't even need to ingest a log to IPFS, you can use any distributed key-value store to store your state etc... the sexy thing to me is that IPFSCluster is more like an add-on to IPFS which adds a feature to the IPFS stack, rather than a separate component based on the traditional tcp/ip stack. On that sense, a Libp2p service sounds good to me (since it's what IPFS uses already).

> on making replication pluggable. Although lets be careful about attaching metadata per hash, that might get to be a significant overhead. I think adding metadata per 'pinset' might be a better tactic there.

Agree, sort of. If a 'pinset' is a set of IPFS hashes which are pinned recursively (so all the subtrees starting from them), and you don't individually track every hash of that tree, you don't have a way to know if parts of the tree have already been persisted because they are shared by a different pinset. So 'pinset' is useful to summarize what needs to be pinned, but still probably need to keep detailed info on each individual hash. Still, metadata is expensive and should be reduced.

Cheers :)
